File: backend/app/models/mantenimiento.py
from datetime import date, datetime, timezone
from decimal import Decimal
from enum import Enum
from typing import Optional

from sqlmodel import Field, SQLModel


class EstadoMantenimiento(str, Enum):
    solicitud  = "solicitud"
    evaluacion = "evaluacion"
    programado = "programado"
    ejecucion  = "ejecucion"
    completado = "completado"
    cerrado    = "cerrado"
    cancelado  = "cancelado"


class ClasificacionMantenimiento(str, Enum):
    preventivo = "preventivo"
    correctivo = "correctivo"


class ModalidadMantenimiento(str, Enum):
    interno = "interno"
    externo = "externo"


class OrigenMantenimiento(str, Enum):
    intranet               = "intranet"
    qr                     = "qr"
    whatsapp               = "whatsapp"
    telefonico_retroactivo = "telefonico_retroactivo"


class PrioridadMantenimiento(str, Enum):
    baja    = "baja"
    media   = "media"
    alta    = "alta"
    urgente = "urgente"


class SolicitudMantenimiento(SQLModel, table=True):
    __tablename__ = "mnt_solicitudes"

    id:          Optional[int] = Field(default=None, primary_key=True)
    consecutivo: str = Field(max_length=30, unique=True)

    titulo:      str = Field(max_length=300)
    descripcion: str

    tipo_mantenimiento: str = Field(max_length=100)
    clasificacion:      str = Field(max_length=20)
    modalidad:          str = Field(max_length=20)

    fecha_proxima_mantenimiento: Optional[date] = Field(default=None)

    estado:           str = Field(default=EstadoMantenimiento.solicitud, max_length=30)
    fecha_programada: Optional[datetime] = Field(default=None)
    notas_evaluacion: Optional[str] = Field(default=None)

    solicitante_id: int
    asignado_id:    Optional[int] = Field(default=None)
    empresa_nombre: Optional[str] = Field(default=None, max_length=100)

    # ── Campos Fase 1 ─────────────────────────────────────────────────────────
    origen:         str = Field(default=OrigenMantenimiento.intranet, max_length=30)
    prioridad:      str = Field(default=PrioridadMantenimiento.media, max_length=20)
    monto_estimado: Optional[Decimal] = Field(default=None)
    monto_real:     Optional[Decimal] = Field(default=None)
    evidencia_url:  Optional[str] = Field(default=None, max_length=500)
    activo_qr_id:   Optional[int] = Field(default=None)

    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class TipoMantenimientoConfig(SQLModel, table=True):
    __tablename__ = "mnt_tipos_config"

    id:     Optional[int] = Field(default=None, primary_key=True)
    nombre: str = Field(max_length=100, unique=True)
    activo: bool = Field(default=True)
    orden:  int = Field(default=0)

    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class HistorialMantenimiento(SQLModel, table=True):
    __tablename__ = "mnt_historial"

    id:              Optional[int] = Field(default=None, primary_key=True)
    solicitud_id:    int = Field(index=True)
    estado_anterior: Optional[str] = Field(default=None, max_length=30)
    estado_nuevo:    str = Field(max_length=30)
    nota:            Optional[str] = Field(default=None)
    usuario_id:      int
    usuario_nombre:  str = Field(max_length=200)
    fecha:           datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class MntAprobacion(SQLModel, table=True):
    """Aprobaciones requeridas para mantenimientos con monto > $2.000.000."""
    __tablename__ = "mnt_aprobaciones"

    id:               Optional[int] = Field(default=None, primary_key=True)
    solicitud_id:     int = Field(index=True)
    aprobador_id:     int
    aprobador_nombre: str = Field(max_length=200)
    rol_aprobador:    str = Field(max_length=50)
    aprobado:         bool = Field(default=True)
    nota:             Optional[str] = Field(default=None)
    fecha:            datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class MntActivoQR(SQLModel, table=True):
    """Activos físicos con código QR para reporte rápido de novedades."""
    __tablename__ = "mnt_activos_qr"

    id:          Optional[int] = Field(default=None, primary_key=True)
    nombre:      str = Field(max_length=200)
    ubicacion:   str = Field(max_length=300)
    descripcion: Optional[str] = Field(default=None)
    activo:      bool = Field(default=True)
    qr_token:    str = Field(max_length=64, unique=True)

    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

File: frontend/src/types/mantenimiento.ts
// Enums
export type EstadoMantenimiento =
  | "solicitud" | "evaluacion" | "programado"
  | "ejecucion" | "completado" | "cerrado" | "cancelado"

export type ClasificacionMantenimiento = "preventivo" | "correctivo"
export type ModalidadMantenimiento     = "interno" | "externo"
export type OrigenMantenimiento        = "intranet" | "qr" | "whatsapp" | "telefonico_retroactivo"
export type PrioridadMantenimiento     = "baja" | "media" | "alta" | "urgente"

// Modelo principal
export interface SolicitudMantenimiento {
  id:                          number
  consecutivo:                 string
  titulo:                      string
  descripcion:                 string
  tipo_mantenimiento:          string
  clasificacion:               ClasificacionMantenimiento
  modalidad:                   ModalidadMantenimiento
  fecha_proxima_mantenimiento: string | null
  estado:                      EstadoMantenimiento
  fecha_programada:            string | null
  notas_evaluacion:            string | null
  solicitante_id:              number
  solicitante_nombre:          string | null
  asignado_id:                 number | null
  asignado_nombre:             string | null
  empresa_nombre:              string | null
  // Campos Fase 1
  origen:              OrigenMantenimiento
  prioridad:           PrioridadMantenimiento
  monto_estimado:      number | null
  monto_real:          number | null
  evidencia_url:       string | null
  activo_qr_id:        number | null
  requiere_aprobacion: boolean
  aprobaciones_count:  number
  created_at:          string
  updated_at:          string
}

export interface SolicitudesMantenimientoListResponse {
  items:  SolicitudMantenimiento[]
  total:  number
  page:   number
  pages:  number
}

export interface TipoMantenimientoConfig {
  id:     number
  nombre: string
  activo: boolean
  orden:  number
}

export interface HistorialMantenimientoEntrada {
  id:              number
  estado_anterior: EstadoMantenimiento | null
  estado_nuevo:    EstadoMantenimiento
  nota:            string | null
  usuario_id:      number
  usuario_nombre:  string
  fecha:           string
}

export interface OCVinculada {
  id:              string
  consecutivo_os:  string
  descripcion:     string
  estado:          string
  nivel_prioridad: string
  fecha_solicitud: string
}

export interface Aprobacion {
  id:               number
  solicitud_id:     number
  aprobador_nombre: string
  rol_aprobador:    "dir_administrativa" | "gerencia_operaciones" | "gerencia_general"
  aprobado:         boolean
  nota:             string | null
  fecha:            string
}

export interface MobileOut {
  solicitud_id:       number
  consecutivo:        string
  titulo:             string
  descripcion:        string
  estado:             string
  asignado_nombre:    string | null
  solicitante_nombre: string | null
}

export interface KpisMes {
  total:            number
  cerradas:         number
  en_curso:         number
  canceladas:       number
  informales:       number
  gasto_total:      number
  gasto_preventivo: number
  gasto_correctivo: number
  gasto_interno:    number
  gasto_externo:    number
}

export interface KpisOut {
  mes_actual:            KpisMes
  por_origen:            Record<string, number>
  pendientes_aprobacion: number
}

export interface AccesoMovilOut {
  url_qr:           string
  url_jwt:          string
  whatsapp_numero:  string | null
  mensaje_whatsapp: string
}

export interface MntNotificacionesConfig {
  whatsapp_numero_default: string
}

// Payloads
export interface CrearMantenimientoPayload {
  titulo:                      string
  descripcion:                 string
  tipo_mantenimiento:          string
  clasificacion:               ClasificacionMantenimiento
  modalidad:                   ModalidadMantenimiento
  fecha_proxima_mantenimiento: string | null
  origen?:                     OrigenMantenimiento
  prioridad?:                  PrioridadMantenimiento
  monto_estimado?:             number | null
  activo_qr_id?:               number | null
}

export interface CrearRetroactivoPayload {
  titulo:             string
  descripcion:        string
  tipo_mantenimiento: string
  clasificacion:      ClasificacionMantenimiento
  modalidad:          ModalidadMantenimiento
  nota_cierre:        string
  monto_real?:        number | null
  evidencia_url?:     string | null
  asignado_id?:       number | null
}

export interface SubirEvidenciaPayload {
  evidencia_url: string
  monto_real?:   number | null
  nota?:         string
}

export interface AprobacionPayload {
  rol_aprobador: "dir_administrativa" | "gerencia_operaciones" | "gerencia_general"
  nota?:         string
}

export interface CambiarEstadoMantenimientoPayload {
  estado_nuevo: EstadoMantenimiento
  nota?:        string
}

export interface CrearOCVinculadaPayload {
  descripcion:                string
  categoria?:                 string
  grupo_articulos?:           string
  nivel_prioridad:            string
  sede?:                      string
  observaciones_solicitante?: string
}

export interface MantenimientoFilters {
  estado?:        EstadoMantenimiento | ""
  clasificacion?: ClasificacionMantenimiento | ""
  modalidad?:     ModalidadMantenimiento | ""
  q?:             string
}
